functions/src/index.ts

```diff
@@ -46,163 +45,194 @@ const genericErrorHandler = (err: any, req: express.Request, res: express.Respon
+    throw new Error("Missing token in headers, query or cookie");
+  } else if (token.startsWith(ReadWriteTokenPrefix)) {
+    return token;
+  } else {
+    return getClaimsFromJWT(token);
+  }
 };
 
 app.use(customHandlers);
 app.use(genericErrorHandler);
 app.use(bodyParser.json());
 app.use(bodyParser.urlencoded({ extended: false }));
 app.use(cors());
-app.use(checkAuth);
-
-const checkEnv = (req: express.Request, res: express.Response, next: express.NextFunction) => {
-  const { env } = req.query;
-  if (!env) {
-    res.error(403, "Missing env query parameter!");
-  }
-  else {
-    req.env = env as string;
-    next();
-  }
-};
 app.use(checkEnv);
 
 app.get('/api/v1/resources', (req, res) => {
-  return BaseResourceObject.FindAll(db, req.env, req.claims, req.query)
-    .then(resources => resources.map((resource) => resource.apiResult()))
-    .then(apiResults => res.success(apiResults))
-    .catch(error => res.error(400, error))
-})
+  try {
+    const claims = authUsingJWT(req);
+    BaseResourceObject.FindAll(db, req.env, claims, req.query)
+      .then(resources => resources.map((resource) => resource.apiResult()))
+      .then(apiResults => res.success(apiResults))
+      .catch(error => res.error(400, error))
+  } catch (error) {
+    res.error(403, error)
+  }
+});
 
 app.get('/api/v1/resources/:id', (req, res) => {
-  return BaseResourceObject.Find(db, req.env, req.params.id)
+  BaseResourceObject.Find(db, req.env, req.params.id)
     .then(resource => res.success(resource.apiResult()))
     .catch(error => res.error(404, error))
-})
+});
 
 app.post('/api/v1/resources', (req, res) => {
-  return BaseResourceObject.Create(db, req.env, req.claims, req.body)
-    .then(resource => res.success(resource.apiResult(), 201))
-    .catch(error => res.error(400, error))
-})
+  // Note that this is the only route where auth is not called directly. Resource can be created anonymously.
+  const token = getToken(req);
+  if (token) {
+    try {
+      const claims = getClaimsFromJWT(token);
+      BaseResourceObject.Create(db, req.env, claims, req.body)
+        .then(resource => res.success(resource.apiResult(), 201))
+        .catch(error => res.error(400, error));
+    } catch (error) {
+      res.error(403, error);
+    }
+  } else {
+    // Anonymous usage based on readWriteToken. NOT RECOMMENDED apart from testing. readWriteTokens are meant to
+    // be used mostly/only to support converted Document Store documents.
+    BaseResourceObject.Create(db, req.env, undefined, req.body)
+      .then(resource => res.success(resource.apiResult(), 201))
+      .catch(error => res.error(400, error))
+  }
+});
 
 app.patch('/api/v1/resources/:id', (req, res) => {
-  return BaseResourceObject.Update(db, req.env, req.claims, req.params.id, req.body)
-    .then(resource => res.success(resource.apiResult()))
-    .catch(error => res.error(400, error))
-})
+  try {
+    const claimsOrReadWriteToken = authUsingJWTOrReadWriteToken(req);
```

Comment:
I think it'd be safer to only allow JWT access to do a patch like this.
Otherwise, I think someone with a readWriteToken would be able to update the authRules which means they could mess up the readWrite token and lock themselves out.

Reply:
I can change it. But the same can happen with JWT. A resource owner can remove access for himself. ;)

Reply:
OK, that's restricted now.